File: 2025_go/04_printing.py
#!/usr/bin/env python3

NEIGHBORS = [(dy, dx) for dy in range(-1, 2) for dx in range(-1, 2) if (dy, dx) != (0, 0)]


def solve04(grid: list[list[bool]], part_b: bool) -> int:
    def count_neighbors(y: int, x: int) -> int:
        return sum(
            1 for dy, dx in NEIGHBORS
            if 0 <= (ny := y + dy) < len(grid)
            and 0 <= (nx := x + dx) < len(grid[ny])
            and grid[ny][nx]
        )

    total = 0
    changed = True
    while changed:
        changed = False
        for y, row in enumerate(grid):
            for x, is_active in enumerate(row):
                if is_active and count_neighbors(y, x) < 4:
                    total += 1
                    if part_b:
                        grid[y][x] = False
                        changed = True
    return total


if __name__ == "__main__":
    with open("inputs/04_sample.txt", 'r') as f:
        s = [[c == '@' for c in line.strip()] for line in f]
    with open("inputs/04_input.txt", 'r') as f:
        i = [[c == '@' for c in line.strip()] for line in f]

    print(f"Sample A: {solve04(s, False)}") # 13
    print(f"A:        {solve04(i, False)}") # 1489
    print(f"Sample B: {solve04(s, True)}")  # 43
    print(f"B:        {solve04(i, True)}")  # 8890

File: 2025_go/04_printing.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	fmt.Println("Sample A", solve04("inputs/04_sample.txt", false)) // 13
	fmt.Println("A", solve04("inputs/04_input.txt", false))         // 1489
	fmt.Println("Sample B", solve04("inputs/04_sample.txt", true))  // 43
	fmt.Println("B", solve04("inputs/04_input.txt", true))          // 8890
}

func solve04(filename string, b bool) int {
	var (
		f, _    = os.Open(filename)
		scanner = bufio.NewScanner(f)
		grid    [][]bool
		total   = 0
		changed = true
	)
	defer f.Close()
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			continue
		}
		grid = append(grid, make([]bool, len(line)))
		for i, c := range line {
			grid[len(grid)-1][i] = c == '@'
		}
	}
	for changed {
		changed = false
		for y := range grid {
			for x := range grid[y] {
				if !grid[y][x] {
					continue
				}
				neighbors := 0
				for yinc := -1; yinc <= 1; yinc++ {
					for xinc := -1; xinc <= 1; xinc++ {
						if (yinc != 0 || xinc != 0) && y+yinc >= 0 && y+yinc < len(grid) && x+xinc >= 0 && x+xinc < len(grid[y+yinc]) && grid[y+yinc][x+xinc] {
							neighbors++
						}
					}
				}
				if neighbors < 4 {
					total++
					if b {
						grid[y][x] = false
						changed = true
					}
				}
			}
		}
	}
	return total
}

func solve04Line(line string) int {
	return 0
}
